src/GitHub.Api/UI/TreeBase.cs

```diff
@@ -332,8 +300,34 @@ private List<TNode> GetLeafNodes(TNode node, int idx)
             return results;
         }
 
+        private void SetCheckStateOnNode(TNode node, bool setChecked)
+        {
+            SetCheckStateOnNode(node, setChecked ? CheckState.Checked : CheckState.Empty);
+        }
 
-        private void ToggleParentFoldersChecked(int idx, TNode node, bool isChecked)
+        private void SetCheckStateOnNode(TNode node, CheckState setCheckState)
+        {
+            var isChecked = setCheckState == CheckState.Checked
+                || setCheckState == CheckState.Mixed;
+
+            var wasChecked = node.CheckState == CheckState.Checked;
+
+            node.CheckState = setCheckState;
+
+            if (!node.IsFolder)
+            {
+                if (isChecked && !wasChecked)
+                {
+                    AddCheckedNode(node);
```

Comment:
Out of curiosity, what do `AddCheckedNode` and `RemoveCheckedNode` do? Are they callbacks—like, "this node is being added/removed to/from the set of checked nodes?"